README.md
<div align="center">
  <br>
  <h1>🍕 trending.khulnasoft.com 🍕</h1>
  <strong>The site that recommends the hottest sauce. Inspired by <a href="https://daily.dev/">daily.dev</a></strong>
</div>
<br>
<p align="center">
  <a href="https://github.com/khulnasoft-opensource/trending/actions/workflows/release.yml">
    <img src="https://github.com/khulnasoft-opensource/trending/actions/workflows/release.yml/badge.svg" alt="Release" style="max-width: 100%;">
  </a>
  <a href="https://github.com/khulnasoft-opensource/trending/actions/workflows/compliance.yml">
    <img src="https://github.com/khulnasoft-opensource/trending/actions/workflows/compliance.yml/badge.svg" alt="Compliance" style="max-width: 100%;">
  </a>
  <a href="https://github.com/khulnasoft-opensource/trending/actions/workflows/codeql-analysis.yml">
    <img src="https://github.com/khulnasoft-opensource/trending/actions/workflows/codeql-analysis.yml/badge.svg" alt="CodeQL" style="max-width: 100%;">
  </a>
  <img src="https://img.shields.io/badge/Dependabot-active-brightgreen.svg" alt="Dependabot Badge">
  <img src="https://img.shields.io/github/languages/code-size/khulnasoft-opensource/trending" alt="GitHub code size in bytes">
  <img src="https://img.shields.io/github/commit-activity/w/khulnasoft-opensource/trending" alt="GitHub commit activity">
  <a href="https://github.com/khulnasoft-opensource/trending/issues">
    <img src="https://img.shields.io/github/issues/khulnasoft-opensource/trending" alt="GitHub issues">
  </a>
  <a href="https://github.com/khulnasoft-opensource/trending/releases">
    <img src="https://img.shields.io/github/v/release/khulnasoft-opensource/trending.svg?style=flat" alt="GitHub Release">
  </a>
  <a href="https://discord.gg/U2peSNf23P">
    <img src="https://img.shields.io/discord/714698561081704529.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2" alt="Discord">
  </a>
  <a href="https://twitter.com/khulnasoft">
    <img src="https://img.shields.io/twitter/follow/khulnasoft?label=Follow&style=social" alt="Twitter">
  </a>
</p>

## Prerequisites

In order to run the project we need the following software binaries installed on our development machines:

- `node>=16.7.0`
- `npm>=8.0.0`

We can also use one of the listed cloud providers we support:

## 🖥️ Local development

To install the application:

```shell
npm ci
```

To start a local copy of the app on port `3000`:

```shell
npm start
```

### 🧪 Test

For running the test suite, use the following command. Since the tests run in watch mode by default, some users may encounter errors about too many files being open. In this case, it may be beneficial to [install watchman](https://facebook.github.io/watchman/docs/install.html).

```shell
npm test
```

You can request a coverage report by running the following command:

```shell
npm run test:coverage
```

For writing tests, the rule is move business or service logic to the lib folder and write unit tests. Logic that needs to be in a React component, then leverage tools like [Cypress](https://www.cypress.io/) or [Vitest mocking](https://vitest.dev/guide/mocking.html) to write tests.

### 📦 Docker builds

A development preview can also be run from docker:

```shell
docker build -t khulnasoft-opensource-hot .
docker run -p 8080:80 khulnasoft-opensource-hot
```

Alternatively you can pull the production container and skip all builds:

```shell
docker run -dit -p 8080:80 ghcr.io/khulnasoft-opensource/trending
```

### 🎨 Code linting

To check the code and styles quality, use the following command:

```shell
npm run lint
```

This will also display during development, but not break on errors.

To fix the linting errors, use the following command:

```shell
npm run format
```

It is advised to run this command before committing or opening a pull request.

### 📕 Types

We have a couple of scripts to check and adjust missing types.

In order to dry run what types would be added to `package.json`:

```shell
npm run types:auto-check
```

In order to add any missing types to `package.json`:

```shell
npm run types:auto-add
```

### 🚀 Production deployment

A production deployment is a complete build of the project, including the build of the static assets.

```shell
npm run build
```

## 🤝 Contributing

We encourage you to contribute to Khulna Soft! Please check out the [Contributing guide](https://docs.khulnasoft.com/contributing/introduction-to-contributing/) for guidelines about how to proceed.

We have a commit utility called [@khulnasoft-opensource/conventional-commit](https://github.com/khulnasoft-opensource/conventional-commit) that helps you write your commits in a way that is easy to understand and process by others.

It is generally integrated as an `npm` script but you can run it with `npx` as well:

```shell
npm run push
```

For any other npm based project or dotnpmrc defaulting to `--yes`:

```shell
npx -y @khulnasoft-opensource/conventional-commit
```

## 🍕 Community

Got Questions? Join the conversation in our [Discord](https://discord.gg/U2peSNf23P).  
Find Khulna Soft videos and release overviews on our [YouTube Channel](https://www.youtube.com/channel/UCklWxKrTti61ZCROE1e5-MQ).

## 🎦 Repository Visualization

[![Visualization of this repository](./public/diagram.svg)
](./src)

## ⚖️ LICENSE

MIT © [Khulna Soft](LICENSE)


CHANGELOG.md
# 📦 khulnasoft-opensource/trending changelog

[![conventional commits](https://img.shields.io/badge/conventional%20commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![semantic versioning](https://img.shields.io/badge/semantic%20versioning-2.0.0-green.svg)](https://semver.org)

> All notable changes to this project will be documented in this file

## [2.33.2-beta.1](https://github.com/khulnasoft-opensource/trending/compare/v2.33.1...v2.33.2-beta.1) (2023-08-22)


### 🐛 Bug Fixes

* fixed nav menu button ([#483](https://github.com/khulnasoft-opensource/trending/issues/483)) ([dba7890](https://github.com/khulnasoft-opensource/trending/commit/dba789099c3bc0cb847c9d927d39d4cbd1da0147))

## [2.33.1](https://github.com/khulnasoft-opensource/trending/compare/v2.33.0...v2.33.1) (2023-08-11)


### 🐛 Bug Fixes

* Footer position on initial render ([#474](https://github.com/khulnasoft-opensource/trending/issues/474)) ([676477d](https://github.com/khulnasoft-opensource/trending/commit/676477dcc71a7acf496c2c0c1bbe9042a4f56d76))
* Load more button visibility issue  and now loads data upon clicking ([#478](https://github.com/khulnasoft-opensource/trending/issues/478)) ([43f1fb0](https://github.com/khulnasoft-opensource/trending/commit/43f1fb0e0ff6dbfc29dce5e8fd4d3d0569a7e72e))
* Styled input field ([#475](https://github.com/khulnasoft-opensource/trending/issues/475)) ([4f2d743](https://github.com/khulnasoft-opensource/trending/commit/4f2d743ad9ada77ce7c42bbb68ceb174b4764f3a))

## [2.33.1-beta.3](https://github.com/khulnasoft-opensource/trending/compare/v2.33.1-beta.2...v2.33.1-beta.3) (2023-08-11)


### 🐛 Bug Fixes

* Load more button visibility issue  and now loads data upon clicking ([#478](https://github.com/khulnasoft-opensource/trending/issues/478)) ([43f1fb0](https://github.com/khulnasoft-opensource/trending/commit/43f1fb0e0ff6dbfc29dce5e8fd4d3d0569a7e72e))

## [2.33.1-beta.2](https://github.com/khulnasoft-opensource/trending/compare/v2.33.1-beta.1...v2.33.1-beta.2) (2023-07-18)


### 🐛 Bug Fixes

* Footer position on initial render ([#474](https://github.com/khulnasoft-opensource/trending/issues/474)) ([676477d](https://github.com/khulnasoft-opensource/trending/commit/676477dcc71a7acf496c2c0c1bbe9042a4f56d76))

## [2.33.1-beta.1](https://github.com/khulnasoft-opensource/trending/compare/v2.33.0...v2.33.1-beta.1) (2023-07-15)


### 🐛 Bug Fixes

* Styled input field ([#475](https://github.com/khulnasoft-opensource/trending/issues/475)) ([4f2d743](https://github.com/khulnasoft-opensource/trending/commit/4f2d743ad9ada77ce7c42bbb68ceb174b4764f3a))

## [2.33.0](https://github.com/khulnasoft-opensource/trending/compare/v2.32.0...v2.33.0) (2023-05-25)


### 🍕 Features

* Remove khulnasoft-opensource/khulnasoft-opensource workflows ([#470](https://github.com/khulnasoft-opensource/trending/issues/470)) ([4ab3e53](https://github.com/khulnasoft-opensource/trending/commit/4ab3e53792d1b672f1f5eeb5808c621e271e546e))


### 🐛 Bug Fixes

* add workflow_call to triage.yml ([3a1bc14](https://github.com/khulnasoft-opensource/trending/commit/3a1bc1416588f4febc4cc219e44e6af48ec95aef))

## [2.33.0-beta.2](https://github.com/khulnasoft-opensource/trending/compare/v2.33.0-beta.1...v2.33.0-beta.2) (2023-05-25)


### 🐛 Bug Fixes

* add workflow_call to triage.yml ([3a1bc14](https://github.com/khulnasoft-opensource/trending/commit/3a1bc1416588f4febc4cc219e44e6af48ec95aef))

## [2.33.0-beta.1](https://github.com/khulnasoft-opensource/trending/compare/v2.32.0...v2.33.0-beta.1) (2023-05-24)


### 🍕 Features

* Remove khulnasoft-opensource/khulnasoft-opensource workflows ([#470](https://github.com/khulnasoft-opensource/trending/issues/470)) ([4ab3e53](https://github.com/khulnasoft-opensource/trending/commit/4ab3e53792d1b672f1f5eeb5808c621e271e546e))

## [2.32.0](https://github.com/khulnasoft-opensource/trending/compare/v2.31.0...v2.32.0) (2023-05-21)


### 🐛 Bug Fixes

* correct squoosh leveled dependencies ([#468](https://github.com/khulnasoft-opensource/trending/issues/468)) ([c6da752](https://github.com/khulnasoft-opensource/trending/commit/c6da7523180ed6789bd9ae2189e152cc7faafa0a))


### 🔁 Continuous Integration

* correct build environment ([7b6ee33](https://github.com/khulnasoft-opensource/trending/commit/7b6ee335256a3f78158b6be5a33e395e92a50518))


### 🤖 Build System

* correct node version in docker build ([ccd2f6c](https://github.com/khulnasoft-opensource/trending/commit/ccd2f6c8da01d1827f8a42423d3212f1e5f42ff2))


### 🍕 Features

* Add flex wrapper around repo list content ([#443](https://github.com/khulnasoft-opensource/trending/issues/443)) ([13f61f9](https://github.com/khulnasoft-opensource/trending/commit/13f61f9b341ae0a536e412cf8ae23d8d825f18fb))

## [2.32.0-beta.1](https://github.com/khulnasoft-opensource/trending/compare/v2.31.1-beta.1...v2.32.0-beta.1) (2023-05-21)


### 🍕 Features

* Add flex wrapper around repo list content ([#443](https://github.com/khulnasoft-opensource/trending/issues/443)) ([13f61f9](https://github.com/khulnasoft-opensource/trending/commit/13f61f9b341ae0a536e412cf8ae23d8d825f18fb))

## [2.31.1-beta.1](https://github.com/khulnasoft-opensource/trending/compare/v2.31.0...v2.31.1-beta.1) (2023-05-18)


### 🐛 Bug Fixes

* correct squoosh leveled dependencies ([#468](https://github.com/khulnasoft-opensource/trending/issues/468)) ([c6da752](https://github.com/khulnasoft-opensource/trending/commit/c6da7523180ed6789bd9ae2189e152cc7faafa0a))


### 🔁 Continuous Integration

* correct build environment ([7b6ee33](https://github.com/khulnasoft-opensource/trending/commit/7b6ee335256a3f78158b6be5a33e395e92a50518))


### 🤖 Build System

* correct node version in docker build ([ccd2f6c](https://github.com/khulnasoft-opensource/trending/commit/ccd2f6c8da01d1827f8a42423d3212f1e5f42ff2))

## [2.31.0](https://github.com/khulnasoft-opensource/trending/compare/v2.30.0...v2.31.0) (2023-05-16)


### 🧑‍💻 Code Refactoring


src/lib/changeCapilization.test.ts
import camelCaseToTitleCase from "./camelCaseToTitleCase";
import { describe, expect, test } from "vitest";

describe("Test: camelCaseToTitleCase()", () => {
  test("open should capitalize", () => {
    const str = "khulnaSoft";
    const result = camelCaseToTitleCase(str);

    expect(result).toBe("Khulna Soft");
  });
});


vite.config.ts
import { defineConfig } from 'vitest/config'
import { createHtmlPlugin } from 'vite-plugin-html'
import { execaSync } from 'execa'
import ViteEslint from '@nabla/vite-plugin-eslint'
import ViteInspect from 'vite-plugin-inspect'
import ViteLegacy from '@vitejs/plugin-legacy'
import ViteReact from '@vitejs/plugin-react'
import ViteVisualizer from 'rollup-plugin-visualizer'
import { VitePluginFonts } from 'vite-plugin-fonts'

import type { ConfigEnv, UserConfig } from 'vite'

// https://vitejs.dev/config/
export default defineConfig(({command, mode}: ConfigEnv): UserConfig => {
  // figure out commands
  const isBuild = command === 'build';

  // figure out modes
  const isDev = mode === "development";
  const isProd = mode === "production";
  const isReport = mode === "report";

  // figure out custom build options
  const isTest = process.env.NODE_ENV === 'test';
  const isLegacy = process.env.VITE_LEGACY || false;
  const isGitpodBuild = process.env.GITPOD_WORKSPACE_URL || false;
  const isReplitBuild = process.env.REPL_SLUG || false;
  const isStackblitzBuild = process.env.STACKBLITZ_ENV || false;
  const isCodeSandboxBuild = process.env.CODESANDBOX_SSE || false;
  const isGlitchBuild = process.env.PROJECT_REMIX_CHAIN || false;
  const isCloudIdeBuild = isGitpodBuild || isReplitBuild || isStackblitzBuild || isCodeSandboxBuild || isGlitchBuild;
  const isNetlifyBuild = process.env.NETLIFY || false;
  const isNetlifyPreviewBuild = isNetlifyBuild && process.env.CONTEXT == 'deploy-preview';
  const isNetlifyChannelBuild = isNetlifyBuild && process.env.CHANNEL !== undefined;

  const config:UserConfig = {
    base: "/",
    mode,
    plugins: [],
    publicDir: "public",
    clearScreen: true,
    server: {
      host: true,
      port: 3000,
      strictPort: true,
      open: !isCloudIdeBuild,
    },
    build: {
      outDir: "build",
      assetsDir: "static",
      sourcemap: !isDev,
      rollupOptions: {},
      manifest: false,
    },
    test: {
      include: ["src/**/*.test.ts", "src/**/*.test.tsx"],
      exclude: ['**/e2e-tests/**'],
    },
    preview: {
      port: 3000,
    }
  };

  config.plugins.push(
    VitePluginFonts({
      google: {
        preconnect: true,
        display: 'swap',
        injectTo: 'head-prepend',
        families: [
          'Inter:wght@400;600;900',
          'Lexend:wght@400;600;900',
          'Open Sans:wght@400;600;900'
        ],
      },
    }),
    ViteEslint(),
    ViteInspect(),
    ViteReact({
      fastRefresh: !isTest,
      // Exclude storybook stories
      exclude: /\.stories\.(t|j)sx?$/,
      include: "src/**/*.(t|j)sx?",
    }),
    createHtmlPlugin({
      minify: isProd && isBuild,
      inject: {
        data: {
          title: `Hot Khulna Soft v${process.env.npm_package_version}`,
          date: new Date().toISOString(),
        },
      },
    })
  );

  isBuild && isLegacy && config.plugins.push(
    ViteLegacy({
      targets: [
        'defaults',
        'not IE 11'
      ]
    })
  );

  isReport && config.plugins.push(
    ViteVisualizer({
      filename: "./build/bundle.html",
      open: true,
      gzipSize: true
    })
  );

  // cloud container shared and specific build options
  isCloudIdeBuild && (config.server.hmr = {
    port: 443,
  });

  if (isGitpodBuild) {
    config.base = process.env.GITPOD_WORKSPACE_URL;
  }

  if (isReplitBuild) {
    config.base = `https://${process.env.REPL_SLUG}-${process.env.REPL_OWNER}.repl.co`;
  }

  if (isStackblitzBuild) {
    const { stdout } = execaSync("hostname");
    config.base = `https://${stdout}--${config.server.port}.local.webcontainer.io/`;
  }

  if (isGlitchBuild) {
    config.base = `https://${process.env.PROJECT_DOMAIN}.glitch.me/`;
  }

  if (isCodeSandboxBuild) {
    const [type, sandbox, id] = process.env.HOSTNAME.split('-');
    config.base = `https://${id}.${type}.code${sandbox}.io/`;
  }

  // netlify build configuration (see netlify.toml)
  if (isNetlifyBuild) {
    config.base = `${process.env.URL}/`;

    if (isNetlifyPreviewBuild) {
      config.base = `${process.env.DEPLOY_PRIME_URL}/`;
    } else {
      if (isNetlifyChannelBuild && ['alpha', 'beta'].includes(process.env.CHANNEL)) {
        const {protocol, hostname} = new URL(process.env.URL);
        config.base = `${protocol}//${process.env.CHANNEL}.${hostname}/`;
      }
    }
  }

  return config;
})
